examples/hybrid/bubble2d_invariant_totalenergy.jl

```diff
@@ -0,0 +1,322 @@
+    χuₕ = @. duₕ = hwgrad(hdiv(cuₕ))
+
+    Spaces.weighted_dss!(dρe)
+    Spaces.weighted_dss!(duₕ)
+
+    κ₄ = 100.0 # m^4/s
+    @. dρe = -κ₄ * hwdiv(cρ * hgrad(χe))
+    @. duₕ = -κ₄ * (hwgrad(hdiv(χuₕ)))
+
+    # 1) Mass conservation
+    If2c = Operators.InterpolateF2C()
+    Ic2f = Operators.InterpolateC2F(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
+    cw = If2c.(fw)
+    fuₕ = Ic2f.(cuₕ)
+    cuw = Geometry.Covariant13Vector.(cuₕ) .+ Geometry.Covariant13Vector.(cw)
+
+    dw .= fw .* 0
+
+    # 1.a) horizontal divergence
+    dρ .-= hdiv.(cρ .* (cuw))
+
+    # 1.b) vertical divergence
+    vdivf2c = Operators.DivergenceF2C(
+        top = Operators.SetValue(Geometry.Contravariant3Vector(0.0)),
+        bottom = Operators.SetValue(Geometry.Contravariant3Vector(0.0)),
+    )
+    # we want the total u³ at the boundary to be zero: we can either constrain
+    # both to be zero, or allow one to be non-zero and set the other to be its
+    # negation
+
+    # explicit part
+    dρ .-= vdivf2c.(Ic2f.(cρ .* cuₕ))
+    # implicit part
+    dρ .-= vdivf2c.(Ic2f.(cρ) .* fw)
+
+    # 2) Momentum equation
+
+    # curl term
+    hcurl = Operators.Curl()
+    # effectively a homogeneous Dirichlet condition on u₁ at the boundary
+    vcurlc2f = Operators.CurlC2F(
+        bottom = Operators.SetCurl(Geometry.Contravariant2Vector(0.0)),
+        top = Operators.SetCurl(Geometry.Contravariant2Vector(0.0)),
+    )
+    
+    fω¹ = hcurl.(fw) 
+    fω¹ .+= vcurlc2f.(cuₕ) 
+
+    # cross product
+    # convert to contravariant
+    # these will need to be modified with topography
+    fu¹ =
+        Geometry.Contravariant1Vector.(
+            Geometry.Covariant13Vector.(Ic2f.(cuₕ)),
+        ) 
+    fu³ = Geometry.Contravariant3Vector.(Geometry.Covariant13Vector.(fw))
+    @. dw -= fω¹ × fu¹ # Covariant3Vector on faces
+    @. duₕ -= If2c(fω¹ × fu³)
+
+    ce = @. cρe / cρ
+    cI = @. ce - Φ(z) - (norm(cuw)^2) / 2
+    cT = @. cI / C_v + T_0
+    cp = @. cρ * R_d * cT
+
+    @. duₕ -= hgrad(cp) / cρ
+    vgradc2f = Operators.GradientC2F(
+        bottom = Operators.SetGradient(Geometry.Covariant3Vector(0.0)),
+        top = Operators.SetGradient(Geometry.Covariant3Vector(0.0)),
+    )
+    @. dw -= vgradc2f(cp) / Ic2f(cρ)
+
+    cE = @. (norm(cuw)^2) / 2 + Φ(z)
+    @. duₕ -= hgrad(cE)
+    @. dw -= vgradc2f(cE)
+
+    # 3) potential temperature
+
+    @. dρe -= hdiv(cuw * (cρe + cp))
+    @. dρe -= vdivf2c(fw * Ic2f(cρe + cp))
+    @. dρe -= vdivf2c(Ic2f(cuₕ * (cρe + cp)))
+
+    fcc = Operators.FluxCorrectionC2C(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
+    fcf = Operators.FluxCorrectionF2F(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
+
+    @. dρ += fcc(fw, cρ)
+    @. dρe += fcc(fw, cρe)
+    # Upwind correction applied to all variables 
+    @. duₕ += fcc(fw, cuₕ)
+    @. dw += fcf(fw, fw)
```

Comment:
@simonbyrne is this correct for the curl form equation (upwind correction should be applied to the advection terms)?

Reply:
Related to this: I think the use of total enthalpy (both here and in the horizontal hyperdiffusion terms) is the right form (As opposed to \rho e)

Reply:
I'm not sure I understand why it should be enthalpy in the hyperdiffusion?

Reply:
This is following our discussion on the diffusion of conserved quantities; I've made the change to this file and will follow up with the corresponding changes in the sphere / 2D examples.